// @turnkey/sdk-browser@4.2.0/dist/index.js
'use strict';

var http = require('@turnkey/http');
var apiKeyStamper = require('@turnkey/api-key-stamper');
var iframeStamper = require('@turnkey/iframe-stamper');
var webauthnStamper = require('@turnkey/webauthn-stamper');
var sdkClient = require('./sdk-client.js');
var storage = require('./storage.js');
var turnkeyHelpers = require('./turnkey-helpers.js');
var base = require('./__types__/base.js');
var browserClients = require('./__clients__/browser-clients.js');



Object.defineProperty(exports, "TurnkeyActivityError", {
	enumerable: true,
	get: function () { return http.TurnkeyActivityError; }
});
Object.defineProperty(exports, "TurnkeyApi", {
	enumerable: true,
	get: function () { return http.TurnkeyApi; }
});
Object.defineProperty(exports, "TurnkeyRequestError", {
	enumerable: true,
	get: function () { return http.TurnkeyRequestError; }
});
Object.defineProperty(exports, "createActivityPoller", {
	enumerable: true,
	get: function () { return http.createActivityPoller; }
});
Object.defineProperty(exports, "getWebAuthnAttestation", {
	enumerable: true,
	get: function () { return http.getWebAuthnAttestation; }
});
Object.defineProperty(exports, "sealAndStampRequestBody", {
	enumerable: true,
	get: function () { return http.sealAndStampRequestBody; }
});
Object.defineProperty(exports, "ApiKeyStamper", {
	enumerable: true,
	get: function () { return apiKeyStamper.ApiKeyStamper; }
});
Object.defineProperty(exports, "signWithApiKey", {
	enumerable: true,
	get: function () { return apiKeyStamper.signWithApiKey; }
});
Object.defineProperty(exports, "IframeEventType", {
	enumerable: true,
	get: function () { return iframeStamper.IframeEventType; }
});
Object.defineProperty(exports, "IframeStamper", {
	enumerable: true,
	get: function () { return iframeStamper.IframeStamper; }
});
Object.defineProperty(exports, "WebauthnStamper", {
	enumerable: true,
	get: function () { return webauthnStamper.WebauthnStamper; }
});
exports.Turnkey = sdkClient.TurnkeyBrowserSDK;
Object.defineProperty(exports, "StorageKeys", {
	enumerable: true,
	get: function () { return storage.StorageKeys; }
});
exports.getStorageValue = storage.getStorageValue;
exports.setStorageValue = storage.setStorageValue;
exports.DEFAULT_APTOS_ACCOUNTS = turnkeyHelpers.DEFAULT_APTOS_ACCOUNTS;
exports.DEFAULT_BITCOIN_MAINNET_P2PKH_ACCOUNTS = turnkeyHelpers.DEFAULT_BITCOIN_MAINNET_P2PKH_ACCOUNTS;
exports.DEFAULT_BITCOIN_MAINNET_P2SH_ACCOUNTS = turnkeyHelpers.DEFAULT_BITCOIN_MAINNET_P2SH_ACCOUNTS;
exports.DEFAULT_BITCOIN_MAINNET_P2TR_ACCOUNTS = turnkeyHelpers.DEFAULT_BITCOIN_MAINNET_P2TR_ACCOUNTS;
exports.DEFAULT_BITCOIN_MAINNET_P2WPKH_ACCOUNTS = turnkeyHelpers.DEFAULT_BITCOIN_MAINNET_P2WPKH_ACCOUNTS;
exports.DEFAULT_BITCOIN_MAINNET_P2WSH_ACCOUNTS = turnkeyHelpers.DEFAULT_BITCOIN_MAINNET_P2WSH_ACCOUNTS;
exports.DEFAULT_BITCOIN_REGTEST_P2PKH_ACCOUNTS = turnkeyHelpers.DEFAULT_BITCOIN_REGTEST_P2PKH_ACCOUNTS;
exports.DEFAULT_BITCOIN_REGTEST_P2SH_ACCOUNTS = turnkeyHelpers.DEFAULT_BITCOIN_REGTEST_P2SH_ACCOUNTS;
exports.DEFAULT_BITCOIN_REGTEST_P2TR_ACCOUNTS = turnkeyHelpers.DEFAULT_BITCOIN_REGTEST_P2TR_ACCOUNTS;
exports.DEFAULT_BITCOIN_REGTEST_P2WPKH_ACCOUNTS = turnkeyHelpers.DEFAULT_BITCOIN_REGTEST_P2WPKH_ACCOUNTS;
exports.DEFAULT_BITCOIN_REGTEST_P2WSH_ACCOUNTS = turnkeyHelpers.DEFAULT_BITCOIN_REGTEST_P2WSH_ACCOUNTS;
exports.DEFAULT_BITCOIN_SIGNET_P2PKH_ACCOUNTS = turnkeyHelpers.DEFAULT_BITCOIN_SIGNET_P2PKH_ACCOUNTS;
exports.DEFAULT_BITCOIN_SIGNET_P2SH_ACCOUNTS = turnkeyHelpers.DEFAULT_BITCOIN_SIGNET_P2SH_ACCOUNTS;
exports.DEFAULT_BITCOIN_SIGNET_P2TR_ACCOUNTS = turnkeyHelpers.DEFAULT_BITCOIN_SIGNET_P2TR_ACCOUNTS;
exports.DEFAULT_BITCOIN_SIGNET_P2WPKH_ACCOUNTS = turnkeyHelpers.DEFAULT_BITCOIN_SIGNET_P2WPKH_ACCOUNTS;
exports.DEFAULT_BITCOIN_SIGNET_P2WSH_ACCOUNTS = turnkeyHelpers.DEFAULT_BITCOIN_SIGNET_P2WSH_ACCOUNTS;
exports.DEFAULT_BITCOIN_TESTNET_P2PKH_ACCOUNTS = turnkeyHelpers.DEFAULT_BITCOIN_TESTNET_P2PKH_ACCOUNTS;
exports.DEFAULT_BITCOIN_TESTNET_P2SH_ACCOUNTS = turnkeyHelpers.DEFAULT_BITCOIN_TESTNET_P2SH_ACCOUNTS;
exports.DEFAULT_BITCOIN_TESTNET_P2TR_ACCOUNTS = turnkeyHelpers.DEFAULT_BITCOIN_TESTNET_P2TR_ACCOUNTS;
exports.DEFAULT_BITCOIN_TESTNET_P2WPKH_ACCOUNTS = turnkeyHelpers.DEFAULT_BITCOIN_TESTNET_P2WPKH_ACCOUNTS;
exports.DEFAULT_BITCOIN_TESTNET_P2WSH_ACCOUNTS = turnkeyHelpers.DEFAULT_BITCOIN_TESTNET_P2WSH_ACCOUNTS;
exports.DEFAULT_COSMOS_ACCOUNTS = turnkeyHelpers.DEFAULT_COSMOS_ACCOUNTS;
exports.DEFAULT_DOGE_MAINNET_ACCOUNTS = turnkeyHelpers.DEFAULT_DOGE_MAINNET_ACCOUNTS;
exports.DEFAULT_DOGE_TESTNET_ACCOUNTS = turnkeyHelpers.DEFAULT_DOGE_TESTNET_ACCOUNTS;
exports.DEFAULT_ETHEREUM_ACCOUNTS = turnkeyHelpers.DEFAULT_ETHEREUM_ACCOUNTS;
exports.DEFAULT_SEI_ACCOUNTS = turnkeyHelpers.DEFAULT_SEI_ACCOUNTS;
exports.DEFAULT_SOLANA_ACCOUNTS = turnkeyHelpers.DEFAULT_SOLANA_ACCOUNTS;
exports.DEFAULT_SUI_ACCOUNTS = turnkeyHelpers.DEFAULT_SUI_ACCOUNTS;
exports.DEFAULT_TON_V3R2_ACCOUNTS = turnkeyHelpers.DEFAULT_TON_V3R2_ACCOUNTS;
exports.DEFAULT_TON_V4R2_ACCOUNTS = turnkeyHelpers.DEFAULT_TON_V4R2_ACCOUNTS;
exports.DEFAULT_TRON_ACCOUNTS = turnkeyHelpers.DEFAULT_TRON_ACCOUNTS;
exports.DEFAULT_XLM_ACCOUNTS = turnkeyHelpers.DEFAULT_XLM_ACCOUNTS;
exports.defaultAptosAccountAtIndex = turnkeyHelpers.defaultAptosAccountAtIndex;
exports.defaultBitcoinMainnetP2PKHAccountAtIndex = turnkeyHelpers.defaultBitcoinMainnetP2PKHAccountAtIndex;
exports.defaultBitcoinMainnetP2SHAccountAtIndex = turnkeyHelpers.defaultBitcoinMainnetP2SHAccountAtIndex;
exports.defaultBitcoinMainnetP2TRAccountAtIndex = turnkeyHelpers.defaultBitcoinMainnetP2TRAccountAtIndex;
exports.defaultBitcoinMainnetP2WPKHAccountAtIndex = turnkeyHelpers.defaultBitcoinMainnetP2WPKHAccountAtIndex;
exports.defaultBitcoinMainnetP2WSHAccountAtIndex = turnkeyHelpers.defaultBitcoinMainnetP2WSHAccountAtIndex;
exports.defaultBitcoinRegtestP2PKHAccountAtIndex = turnkeyHelpers.defaultBitcoinRegtestP2PKHAccountAtIndex;
exports.defaultBitcoinRegtestP2SHAccountAtIndex = turnkeyHelpers.defaultBitcoinRegtestP2SHAccountAtIndex;
exports.defaultBitcoinRegtestP2TRAccountAtIndex = turnkeyHelpers.defaultBitcoinRegtestP2TRAccountAtIndex;
exports.defaultBitcoinRegtestP2WPKHAccountAtIndex = turnkeyHelpers.defaultBitcoinRegtestP2WPKHAccountAtIndex;
exports.defaultBitcoinRegtestP2WSHAccountAtIndex = turnkeyHelpers.defaultBitcoinRegtestP2WSHAccountAtIndex;
exports.defaultBitcoinSignetP2PKHAccountAtIndex = turnkeyHelpers.defaultBitcoinSignetP2PKHAccountAtIndex;
exports.defaultBitcoinSignetP2SHAccountAtIndex = turnkeyHelpers.defaultBitcoinSignetP2SHAccountAtIndex;
exports.defaultBitcoinSignetP2TRAccountAtIndex = turnkeyHelpers.defaultBitcoinSignetP2TRAccountAtIndex;
exports.defaultBitcoinSignetP2WPKHAccountAtIndex = turnkeyHelpers.defaultBitcoinSignetP2WPKHAccountAtIndex;
exports.defaultBitcoinSignetP2WSHAccountAtIndex = turnkeyHelpers.defaultBitcoinSignetP2WSHAccountAtIndex;
exports.defaultBitcoinTestnetP2PKHAccountAtIndex = turnkeyHelpers.defaultBitcoinTestnetP2PKHAccountAtIndex;
exports.defaultBitcoinTestnetP2SHAccountAtIndex = turnkeyHelpers.defaultBitcoinTestnetP2SHAccountAtIndex;
exports.defaultBitcoinTestnetP2TRAccountAtIndex = turnkeyHelpers.defaultBitcoinTestnetP2TRAccountAtIndex;
exports.defaultBitcoinTestnetP2WPKHAccountAtIndex = turnkeyHelpers.defaultBitcoinTestnetP2WPKHAccountAtIndex;
exports.defaultBitcoinTestnetP2WSHAccountAtIndex = turnkeyHelpers.defaultBitcoinTestnetP2WSHAccountAtIndex;
exports.defaultCosmosAccountAtIndex = turnkeyHelpers.defaultCosmosAccountAtIndex;
exports.defaultDogeMainnetAccountAtIndex = turnkeyHelpers.defaultDogeMainnetAccountAtIndex;
exports.defaultDogeTestnetAccountAtIndex = turnkeyHelpers.defaultDogeTestnetAccountAtIndex;
exports.defaultEthereumAccountAtIndex = turnkeyHelpers.defaultEthereumAccountAtIndex;
exports.defaultSeiAccountAtIndex = turnkeyHelpers.defaultSeiAccountAtIndex;
exports.defaultSolanaAccountAtIndex = turnkeyHelpers.defaultSolanaAccountAtIndex;
exports.defaultSuiAccountAtIndex = turnkeyHelpers.defaultSuiAccountAtIndex;
exports.defaultTonV3r2AccountAtIndex = turnkeyHelpers.defaultTonV3r2AccountAtIndex;
exports.defaultTonV4r2AccountAtIndex = turnkeyHelpers.defaultTonV4r2AccountAtIndex;
exports.defaultTronAccountAtIndex = turnkeyHelpers.defaultTronAccountAtIndex;
exports.defaultXlmAccountAtIndex = turnkeyHelpers.defaultXlmAccountAtIndex;
exports.defaultXrpAccountAtIndex = turnkeyHelpers.defaultXrpAccountAtIndex;
Object.defineProperty(exports, "AuthClient", {
	enumerable: true,
	get: function () { return base.AuthClient; }
});
Object.defineProperty(exports, "SessionType", {
	enumerable: true,
	get: function () { return base.SessionType; }
});
exports.TurnkeyBrowserClient = browserClients.TurnkeyBrowserClient;
exports.TurnkeyIframeClient = browserClients.TurnkeyIframeClient;
exports.TurnkeyPasskeyClient = browserClients.TurnkeyPasskeyClient;
exports.TurnkeyWalletClient = browserClients.TurnkeyWalletClient;
