// @turnkey/sdk-browser@4.2.0/dist/sdk-client.mjs
import { WalletStamper } from '@turnkey/wallet-stamper';
import { IframeStamper } from '@turnkey/iframe-stamper';
import { WebauthnStamper } from '@turnkey/webauthn-stamper';
import WindowWrapper from './__polyfills__/window.mjs';
import { TurnkeyRequestError, SessionType } from './__types__/base.mjs';
import { getStorageValue, StorageKeys, removeStorageValue } from './storage.mjs';
import { TurnkeyPasskeyClient, TurnkeyIframeClient, TurnkeyWalletClient, TurnkeyBrowserClient } from './__clients__/browser-clients.mjs';
import { VERSION } from './__generated__/version.mjs';

class TurnkeyBrowserSDK {
    constructor(config) {
        /**
         * Creates a passkey client. The parameters override the default values passed to the underlying Turnkey `WebauthnStamper`
         * @param PasskeyClientParams
         * @returns new TurnkeyPasskeyClient
         */
        this.passkeyClient = (params) => {
            const targetRpId = params?.rpId ?? this.config.rpId ?? WindowWrapper.location.hostname;
            if (!targetRpId) {
                throw new Error("Tried to initialize a passkey client with no rpId defined");
            }
            this.stamper = new WebauthnStamper({
                rpId: targetRpId,
                ...(params?.timeout !== undefined && { timeout: params?.timeout }),
                ...(params?.userVerification !== undefined && {
                    userVerification: params?.userVerification,
                }),
                ...(params?.allowCredentials !== undefined && {
                    allowCredentials: params?.allowCredentials,
                }),
            });
            return new TurnkeyPasskeyClient({
                stamper: this.stamper,
                apiBaseUrl: this.config.apiBaseUrl,
                organizationId: this.config.defaultOrganizationId,
            });
        };
        this.iframeClient = async (params) => {
            if (!params.iframeUrl) {
                throw new Error("Tried to initialize iframeClient with no iframeUrl defined");
            }
            const TurnkeyIframeElementId = params.iframeElementId ?? "turnkey-default-iframe-element-id";
            this.stamper = new IframeStamper({
                iframeContainer: params.iframeContainer,
                iframeUrl: params.iframeUrl,
                iframeElementId: TurnkeyIframeElementId,
            });
            await this.stamper.init(params.dangerouslyOverrideIframeKeyTtl ?? undefined);
            return new TurnkeyIframeClient({
                stamper: this.stamper,
                apiBaseUrl: this.config.apiBaseUrl,
                organizationId: this.config.defaultOrganizationId,
            });
        };
        this.walletClient = (wallet) => {
            return new TurnkeyWalletClient({
                stamper: new WalletStamper(wallet),
                wallet,
                apiBaseUrl: this.config.apiBaseUrl,
                organizationId: this.config.defaultOrganizationId,
            });
        };
        this.serverSign = async (methodName, params, serverSignUrl) => {
            const targetServerSignUrl = serverSignUrl ?? this.config.serverSignUrl;
            if (!targetServerSignUrl) {
                throw new Error("Tried to call serverSign with no serverSignUrl defined");
            }
            const stringifiedBody = JSON.stringify({
                methodName: methodName,
                params: params,
            });
            const response = await fetch(targetServerSignUrl, {
                method: "POST",
                headers: {
                    "Content-Type": "application/json",
                    "X-Client-Version": VERSION,
                },
                body: stringifiedBody,
                redirect: "follow",
            });
            if (!response.ok) {
                let res;
                try {
                    res = await response.json();
                }
                catch (_) {
                    throw new Error(`${response.status} ${response.statusText}`);
                }
                throw new TurnkeyRequestError(res);
            }
            const data = await response.json();
            return data;
        };
        /**
         * If there is a valid, current user session, this will return a read-enabled TurnkeyBrowserClient that can make read requests to Turnkey without additional authentication. This is powered by a session header resulting from a prior successful `login` call.
         *
         * @returns {Promise<TurnkeyBrowserClient | undefined>}
         */
        this.currentUserSession = async () => {
            const currentUser = await this.getCurrentUser();
            if (!currentUser?.session?.read) {
                return;
            }
            if (currentUser?.session?.read?.expiry > Date.now()) {
                return new TurnkeyBrowserClient({
                    readOnlySession: currentUser?.session?.read?.token,
                    apiBaseUrl: this.config.apiBaseUrl,
                    organizationId: currentUser?.organization?.organizationId ??
                        this.config.defaultOrganizationId,
                });
            }
            else {
                await this.logout();
            }
            return;
        };
        /**
         * If there is a valid, current read-session, this will return an auth bundle and its expiration. This auth bundle can be used in conjunction with an iframeStamper to create a read + write session.
         * @deprecated use `getSession` instead
         * @returns {Promise<ReadWriteSession | undefined>}
         */
        this.getReadWriteSession = async () => {
            const currentUser = await getStorageValue(StorageKeys.UserSession);
            if (currentUser?.session?.write) {
                if (currentUser.session.write.expiry > Date.now()) {
                    return currentUser.session.write;
                }
                else {
                    await removeStorageValue(StorageKeys.ReadWriteSession);
                }
            }
            return;
        };
        /**
         * If there is a valid, active READ_WRITE session, this will return it
         *
         * @returns {Promise<Session | undefined>}
         */
        this.getSession = async () => {
            const currentSession = await getStorageValue(StorageKeys.Session);
            if (currentSession?.sessionType === SessionType.READ_WRITE) {
                if (currentSession?.expiry > Date.now()) {
                    return currentSession;
                }
                else {
                    await removeStorageValue(StorageKeys.Session);
                }
            }
            return;
        };
        /**
         * Fetches the current user's organization details.
         *
         * @returns {Promise<SubOrganization | undefined>}
         */
        this.getCurrentSubOrganization = async () => {
            const currentUser = await this.getCurrentUser();
            return currentUser?.organization;
        };
        /**
         * Fetches the currently active user.
         *
         * @returns {Promise<User | undefined>}
         */
        this.getCurrentUser = async () => {
            try {
                const session = await getStorageValue(StorageKeys.Session);
                if (session?.userId && session?.organizationId) {
                    return {
                        userId: session.userId,
                        organization: {
                            organizationId: session.organizationId,
                            organizationName: "",
                        },
                        session: {
                            ...(session.sessionType === SessionType.READ_ONLY && {
                                read: {
                                    token: session.token,
                                    expiry: session.expiry,
                                },
                            }),
                            ...(session.sessionType === SessionType.READ_WRITE && {
                                write: {
                                    credentialBundle: session.token,
                                    expiry: session.expiry,
                                },
                            }),
                        },
                    };
                }
                else {
                    return undefined;
                }
            }
            catch (error) {
                return;
            }
        };
        /**
         * Clears out all data pertaining to an end user session.
         *
         * @returns {Promise<boolean>}
         */
        this.logout = async () => {
            await removeStorageValue(StorageKeys.AuthBundle); // DEPRECATED
            await removeStorageValue(StorageKeys.CurrentUser);
            await removeStorageValue(StorageKeys.UserSession);
            await removeStorageValue(StorageKeys.ReadWriteSession);
            await removeStorageValue(StorageKeys.Client);
            await removeStorageValue(StorageKeys.Session);
            return true;
        };
        this.config = config;
    }
}

export { TurnkeyBrowserClient, TurnkeyBrowserSDK, TurnkeyIframeClient, TurnkeyPasskeyClient, TurnkeyWalletClient };
