// @turnkey/crypto@2.3.1/dist/constants.js
'use strict';

const SUITE_ID_1 = new Uint8Array([75, 69, 77, 0, 16]); //KEM suite ID
const SUITE_ID_2 = new Uint8Array([72, 80, 75, 69, 0, 16, 0, 1, 0, 2]); //HPKE suite ID
const HPKE_VERSION = new Uint8Array([72, 80, 75, 69, 45, 118, 49]); //HPKE-v1
const LABEL_SECRET = new Uint8Array([115, 101, 99, 114, 101, 116]); //secret
const LABEL_EAE_PRK = new Uint8Array([101, 97, 101, 95, 112, 114, 107]); //eae_prk
const LABEL_SHARED_SECRET = new Uint8Array([
    115, 104, 97, 114, 101, 100, 95, 115, 101, 99, 114, 101, 116,
]); //shared_secret
const AES_KEY_INFO = new Uint8Array([
    0, 32, 72, 80, 75, 69, 45, 118, 49, 72, 80, 75, 69, 0, 16, 0, 1, 0, 2, 107,
    101, 121, 0, 143, 195, 174, 184, 50, 73, 10, 75, 90, 179, 228, 32, 35, 40,
    125, 178, 154, 31, 75, 199, 194, 34, 192, 223, 34, 135, 39, 183, 10, 64, 33,
    18, 47, 63, 4, 233, 32, 108, 209, 36, 19, 80, 53, 41, 180, 122, 198, 166, 48,
    185, 46, 196, 207, 125, 35, 69, 8, 208, 175, 151, 113, 201, 158, 80,
]); //key
const IV_INFO = new Uint8Array([
    0, 12, 72, 80, 75, 69, 45, 118, 49, 72, 80, 75, 69, 0, 16, 0, 1, 0, 2, 98, 97,
    115, 101, 95, 110, 111, 110, 99, 101, 0, 143, 195, 174, 184, 50, 73, 10, 75,
    90, 179, 228, 32, 35, 40, 125, 178, 154, 31, 75, 199, 194, 34, 192, 223, 34,
    135, 39, 183, 10, 64, 33, 18, 47, 63, 4, 233, 32, 108, 209, 36, 19, 80, 53,
    41, 180, 122, 198, 166, 48, 185, 46, 196, 207, 125, 35, 69, 8, 208, 175, 151,
    113, 201, 158, 80,
]); //base_nonce
const PRODUCTION_SIGNER_PUBLIC_KEY = "04cf288fe433cc4e1aa0ce1632feac4ea26bf2f5a09dcfe5a42c398e06898710330f0572882f4dbdf0f5304b8fc8703acd69adca9a4bbf7f5d00d20a5e364b2569";

exports.AES_KEY_INFO = AES_KEY_INFO;
exports.HPKE_VERSION = HPKE_VERSION;
exports.IV_INFO = IV_INFO;
exports.LABEL_EAE_PRK = LABEL_EAE_PRK;
exports.LABEL_SECRET = LABEL_SECRET;
exports.LABEL_SHARED_SECRET = LABEL_SHARED_SECRET;
exports.PRODUCTION_SIGNER_PUBLIC_KEY = PRODUCTION_SIGNER_PUBLIC_KEY;
exports.SUITE_ID_1 = SUITE_ID_1;
exports.SUITE_ID_2 = SUITE_ID_2;
